### PR TITLE
{ARM} Fix race condition in bicep unit tests

### PR DESCRIPTION
**Pull request by shenglol**
**Related command**
`az bicep *`

**Description**
~~This main purpose of the PR is to reduce Azure CLI’s packaging maintenance overheads based on @jiasli's feedback https://github.com/Azure/azure-cli/pull/16857#discussion_r1204981875.~~

The PR also fixes race conditions in `test_resource_bicep.py` by having each test uses their own `DummyCli` instance with `random_config_dir` set to `True`.

**Testing Guide**
The existing tests should be sufficient the cover the changes.

**History Notes**

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
ARM

**Comment by shenglol** [comment]
@jiasli @zhoxing-ms Could you take a look?

**Comment by shenglol** [comment]
Well, looks like azure-cli-extensions now has a dependency on `semver` (see https://github.com/Azure/azure-cli-extensions/blob/0b2cbaab7ecd11a8cd99eb5d455045abb22ee9c9/src/fleet/azext_fleet/_validators.py#L5), which causes the failure in CI. Should I add the `semver` dependency back in the requirements files?

**Comment by zhoxing-ms** [comment]
> Well, looks like azure-cli-extensions now has a dependency on semver (see https://github.com/Azure/azure-cli-extensions/blob/0b2cbaab7ecd11a8cd99eb5d455045abb22ee9c9/src/fleet/azext_fleet/_validators.py#L5), which causes the failure in CI. Should I add the semver dependency back in the requirements files?

Okay, this can avoid causing breaking changes to other dependent places

**Comment by shenglol** [comment]
> > Well, looks like azure-cli-extensions now has a dependency on semver (see https://github.com/Azure/azure-cli-extensions/blob/0b2cbaab7ecd11a8cd99eb5d455045abb22ee9c9/src/fleet/azext_fleet/_validators.py#L5), which causes the failure in CI. Should I add the semver dependency back in the requirements files?
> 
> Okay, this can avoid causing breaking changes to other dependent places

Since we cannot remove the `semver` dependency, I restored the _bicep.py to use `semver` instead of `package.versioning`. In the future, if we start to publish pre-releases / private versions for Bicep, the version number may not be compliant with PEP 440.

**Comment by zhoxing-ms** [comment]
@shenglol Actually, I think we can skip deleting B from A in order not to affect other modules/extensions. However, to ensure the consistency in version management, it is still recommended that you use `package.versioning` instead of `semver`

**Comment by shenglol** [comment]
> @shenglol Shenglong Li FTE Actually, I think we can skip removing `semver` from `setup.py` and `requirements.py3.xxx` in order not to affect other modules/extensions. However, to ensure the consistency in version management, it is still recommended that you use `package.versioning` instead of `semver`

To clarify, Bicep CLI is technically using semantic versioning. The version is set by [Nerdbank.GitVersioning](https://github.com/dotnet/Nerdbank.GitVersioning), which produces semver-compatible versions. Currently, for public releases, the version is always in the format of MAJOR.MINOR.PATCH and is also compatible with PEP 440. However, in the future, we might want to publish private or pre-release versions that may contain suffixes (e.g., commit hashes) that are not necessarily compatible with PEP 440.

**Review by zhoxing-ms — APPROVED** [review]

**Comment by zhoxing-ms** [comment]
@jiasli Could you please help review this PR?

**Review by bebound — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/resource/_bicep.py [hunk 1]
```diff
@@ -302,9 +296,9 @@ def _get_bicep_installation_path(system):
     raise ValidationError(f'The platform "{system}" is not supported.')
 
 
-def _extract_semver(text):
+def _extract_version(text):
     semver_match = re.search(_semver_pattern, text)
-    return semver_match.group(0) if semver_match else None
+    return semver.VersionInfo.parse(semver_match.group(0)) if semver_match else None
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/resource/_bicep.py:301]: If `semver` is indeed required, we should consider revive https://github.com/Azure/azure-cli/pull/26523.

File: src/azure-cli/azure/cli/command_modules/resource/tests/latest/test_resource_bicep.py [hunk 2]
```diff
@@ -19,20 +20,21 @@
 from azure.cli.core.azclierror import InvalidTemplateError
 from azure.cli.core.mock import DummyCli
 
-cli_ctx = DummyCli()
-
 
 class TestBicep(unittest.TestCase):
+    def setUp(self):
+        self.cli_ctx = DummyCli(random_config_dir=True)
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/resource/tests/latest/test_resource_bicep.py:26]: This contradicts the convention used in https://github.com/Azure/azure-cli/pull/25689 that `random_config_dir=True` is set in `super().__init__` invocation:

https://github.com/Azure/azure-cli/blob/b223dd9d4379d5c633a9f4dba4941ae791531126/src/azure-cli/azure/cli/command_modules/config/tests/latest/test_config.py#L16-L17

Comment (bebound) [reply src/azure-cli/azure/cli/command_modules/resource/tests/latest/test_resource_bicep.py:26]: Its superclass is `unittest.TestCase`, which does not has `random_config_dir` param.

**Pull request merged**